kernel/src/process/process/mod.rs

```diff
@@ -343,254 +338,274 @@ impl Process {
-            return_errno_with_message!(
-                Errno::EPERM,
-                "process group leader cannot be moved to new session."
-            );
-        }
-
-        let session = self.session().unwrap();
-
-        // Lock order: session table -> group table -> group of process -> group inner -> session inner
+    /// This method will return `EPERM` if an existing process group has the same identifier as the
+    /// process ID. This means that the process is or was a process group leader and that the
+    /// process group is still alive.
+    pub fn to_new_session(self: &Arc<Self>) -> Result<Sid> {
+        // Lock order: session table -> group table -> group of process
+        // -> group inner -> session inner
         let mut session_table_mut = process_table::session_table_mut();
         let mut group_table_mut = process_table::group_table_mut();
-        let mut self_group_mut = self.process_group.lock();
 
         if session_table_mut.contains_key(&self.pid) {
-            return_errno_with_message!(Errno::EPERM, "cannot create new session");
+            // FIXME: According to the Linux implementation, this check should be removed, so we'll
+            // return `EPERM` due to hitting the following check. However, we need to work around a
+            // gVisor bug. The upstream gVisor has fixed the issue in:
+            // <https://github.com/google/gvisor/commit/582f7bf6c0ccccaeb1215a232709df38d5d409f7>.
+            return Ok(self.pid);
         }
-
         if group_table_mut.contains_key(&self.pid) {
-            return_errno_with_message!(Errno::EPERM, "cannot create process group");
+            return_errno_with_message!(
+                Errno::EPERM,
+                "a process group leader cannot be moved to a new session"
+            );
         }
 
-        // Removes the process from old group
-        if let Some(old_group) = self_group_mut.upgrade() {
-            let mut group_inner = old_group.inner.lock();
-            let mut session_inner = session.inner.lock();
-            group_inner.remove_process(&self.pid);
-            *self_group_mut = Weak::new();
-
-            if group_inner.is_empty() {
-                group_table_mut.remove(&old_group.pgid());
-                debug_assert!(session_inner.process_groups.contains_key(&old_group.pgid()));
-                session_inner.process_groups.remove(&old_group.pgid());
-
-                if session_inner.is_empty() {
-                    session_table_mut.remove(&session.sid());
-                }
+        let mut process_group_mut = self.process_group.lock();
+
+        self.clear_old_group(
+            &mut process_group_mut,
+            &mut session_table_mut,
+            &mut group_table_mut,
+        );
+
+        Ok(self.set_new_session(
+            &mut process_group_mut,
+            &mut session_table_mut,
+            &mut group_table_mut,
+        ))
+    }
+
+    pub(super) fn clear_old_group(
```

Comment:
It would be more appropriate to rename this method as `clear_old_group_and_session`. When encountering this method, I was initially puzzled as to why it wasn't invoked in `to_other_group`, but rather only in `set_new_session`. The reason is that `to_other_group` simply transfers the process to a different group within the same session, meaning that the session remains unchanged and thus doesn't require clearing.